### PR TITLE
[RFC 2195] Document new type representations

### LINKED ISSUE #244: Replace remaining "C-like enumerations" with "fieldless enumerations"
See rust-lang/rust#46348. Looks like a few snuck through: https://doc.rust-lang.org/nightly/reference/type-layout.html#reprc-enums, for instance.

### PR DESCRIPTION
**Pull request by Havvy**
This PR does two things:

1. Documents RFC 1240.
2. Splits the Enum sections for the C and Primitive representations into field-less enums and enums with fields.

Incidentally, fixes #244

### CODE REVIEW

**Comment by Havvy** [comment]
As this is Gankro's RFC,

r? @Gankro

File: src/type-layout.md [hunk 1]
```diff
@@ -222,48 +223,165 @@ assert_eq!(std::mem::size_of::<SizeRoundedUp>(), 8);  // Size of 6 from b,
...
 
-It is an error for [zero-variant enumerations] to have the `C` representation.
+#### \#[repr(C)] Enums With Fields
 
-For all other enumerations, the layout is unspecified.
+For enums with fields, the `C` representation is a struct with representation
+`C` of two fields where the first field is a field-less enum with the `C`
+representation that has one variant for each variant in the enum with fields
+and the second field a union with the `C` representation that's fields consist
+of structs with the `C` representation corresponding to each variant in the
```
Comment (alercah) [inline src/type-layout.md]: This is a very long sentence that's hard to parse. I'd recommend splitting the description of the fields into separate sentences, and perhaps rather than writing "with the `C` representation" inline in each of them, just write at the end something like "The structs and unions declared here all have the `C` representation themselves."

File: src/type-layout.md [hunk 2]
```diff
@@ -222,48 +223,165 @@ assert_eq!(std::mem::size_of::<SizeRoundedUp>(), 8);  // Size of 6 from b,
...
 
-For [C-like enumerations], they set the size and alignment to be the same as the
-primitive type of the same name. For example, a C-like enumeration with a `u8`
-representation can only have discriminants between 0 and 255 inclusive.
+Combining the `C` representation and a primitive representation is described
+[above](#c-primitive-representation).
 
-It is an error for [zero-variant enumerations] to have a primitive
-representation.
+#### Primitive Fepresentation of Field-less Enums
```
Comment (alercah) [inline src/type-layout.md]: typo: Fepresentation

File: src/type-layout.md [hunk 3]
```diff
@@ -222,48 +223,165 @@ assert_eq!(std::mem::size_of::<SizeRoundedUp>(), 8);  // Size of 6 from b,
...
-Likewise, combining two primitive representations together is unspecified.
+#### Primitive Representation of Enums With Fields
+
+For enums with fields, the enum will have the same type layout a union with the
+`C` representation that's fields consist of structs with the `C` representation
+corresponding to each variant in the enum. The first field in each struct is
+the same field-less enum with the same primitive representation that is
+the enum with all fields in its variants removed and the rest of the fields
+consisting of the fields of the corresponding variant in the order defined in
+original enumeration.
```
Comment (alercah) [inline src/type-layout.md]: This sentence is not as bad, but it might be possible to clean it up somewhat as well.

File: src/type-layout.md [hunk 4]
```diff
@@ -222,48 +223,165 @@ assert_eq!(std::mem::size_of::<SizeRoundedUp>(), 8);  // Size of 6 from b,
...
+
+Because unions with non-copy fields aren't allowed, this representation can only
+be used if every field is also [`Copy`].
+
+> Note: This is commonly different than what is done in C and C++. Projects in
+> those languages often use a tuple of `(enum, payload)`. For making your enum
+> represented like that, use the `C` representation.
```
Comment (alercah) [inline src/type-layout.md]: Perhaps link up to C representation here?

File: src/type-layout.md [hunk 5]
```diff
@@ -222,48 +223,165 @@ assert_eq!(std::mem::size_of::<SizeRoundedUp>(), 8);  // Size of 6 from b,
...
+
+For enums with fields, the enum will have the same type layout a union with the
+`C` representation that's fields consist of structs with the `C` representation
+corresponding to each variant in the enum. The first field in each struct is
+the same field-less enum with the same primitive representation that is
+the enum with all fields in its variants removed and the rest of the fields
+consisting of the fields of the corresponding variant in the order defined in
+original enumeration.
+
+Because unions with non-copy fields aren't allowed, this representation can only
```
Comment (alercah) [inline src/type-layout.md]: This doesn't appear to be the case: https://play.rust-lang.org/?gist=6d0c155ebdffe17c6a3cd9683318209b&version=nightly

File: src/type-layout.md [hunk 6]
```diff
@@ -222,48 +223,165 @@ assert_eq!(std::mem::size_of::<SizeRoundedUp>(), 8);  // Size of 6 from b,
...
+enum MyEnum {
+    A(u32),
+    B(f32, u64),
+    C { x: u32, y: u8 },
+    D,
+}
+
+// has the same type layout as this union
+#[repr(C)]
+#[derive(Clone, Copy)]
```
Comment (alercah) [inline src/type-layout.md]: I'd omit the `derive` here; it doesn't tell us anything about the layout and isn't always correct.

File: src/type-layout.md (hunk 1 above)
Comment (Havvy) [reply src/type-layout.md]: Yeah, I wasn't happy with it either. I had a personal breakthrough in how I can describe it, so now it's longer and with paragraph breaks. Saying they all have `C` representation on its own line is also a nice win. 👍

File: src/type-layout.md [hunk 7]
```diff
@@ -222,48 +223,178 @@ assert_eq!(std::mem::size_of::<SizeRoundedUp>(), 8);  // Size of 6 from b,
...
+original enumeration.
 
-For all other enumerations, the layout is unspecified.
+Because unions with non-copy fields aren't allowed, this representation can only
+be used if every field is also [`Copy`].
 
-Likewise, combining two primitive representations together is unspecified.
+> Note: This is commonly different than what is done in C and C++. Projects in
+> those languages often use a tuple of `(enum, payload)`. For making your enum
+> represented like that, use the `C` representation.
```
Comment (Gankro) [inline src/type-layout.md]: I'm not sure this is necessary. Or at least I would eliminate the reference to "what's commonly done in C(++)" which is like, never a true statement.

File: src/type-layout.md [hunk 8]
```diff
@@ -222,48 +223,178 @@ assert_eq!(std::mem::size_of::<SizeRoundedUp>(), 8);  // Size of 6 from b,
...
+
+```rust
+// This custom enum
```
Comment (Gankro) [inline src/type-layout.md]: Similar notes on this example as the previous (although you used my preferred naming scheme on this one..?)

File: src/type-layout.md [hunk 9]
```diff
@@ -222,48 +223,178 @@ assert_eq!(std::mem::size_of::<SizeRoundedUp>(), 8);  // Size of 6 from b,
...
+> bitflags in `C`. In contrast, Rust’s field-less enums can only legally hold
+> the discrimnant values, everything else is [undefined behavior]. Therefore,
+> using a field-less enum in FFI to model a C `enum` is often wrong.
 
-It is an error for [zero-variant enumerations] to have the `C` representation.
+#### \#[repr(C)] Enums With Fields
 
-For all other enumerations, the layout is unspecified.
+For enums with fields, the `C` representation is defined to be the same as the
+follow types. These types don't actually exist, so the names are only here to
```
[the remark this answers is not in the document]
Comment (Havvy) [reply src/type-layout.md]: following*

File: src/type-layout.md [hunk 10]
```diff
@@ -222,48 +223,178 @@ assert_eq!(std::mem::size_of::<SizeRoundedUp>(), 8);  // Size of 6 from b,
...
+
+### Primitive Representations
 
 The *primitive representations* are the representations with the same names as
 the primitive integer types. That is: `u8`, `u16`, `u32`, `u64`, `usize`, `i8`,
 `i16`, `i32`, `i64`, and `isize`.
 
-Primitive representations can only be applied to enumerations.
+Primitive representations can only be applied to enumerations, and have
+different behavior whether the enum has fields or no fields. It is an error
```
[the remark this answers is not in the document]
Comment (Havvy) [reply src/type-layout.md]: From a logical standpoint, no. Could be "depending on".

File: src/type-layout.md (hunk 8 above)
Comment (Havvy) [reply src/type-layout.md]: The last commit was a WIP where I was transitioning from this style (that you like better) to the style I used in `#[repr(C)]` (that you like less).

**Comment by alercah** [comment]
@Havvy what's the status of this?

**Comment by Havvy** [comment]
The somewhat conflicting feedback scared me on this, and I still need to actually look at the feedback...

**Comment by Centril** [comment]
Ping @Havvy, did you get a chance to actually look at the feedback?

**Comment by Havvy** [comment]
I don't know if and when I'd get back to this, and it needs a lot of work. So I'm going to just close this.
